src/libsync/propagatecommonzsync.cpp

```diff
@@ -0,0 +1,320 @@
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#include "config.h"
+#include "propagateupload.h"
+#include "owncloudpropagator_p.h"
+#include "networkjobs.h"
+#include "account.h"
+#include "common/syncjournaldb.h"
+#include "common/syncjournalfilerecord.h"
+#include "common/utility.h"
+#include "filesystem.h"
+#include "propagatorjobs.h"
+#include "syncengine.h"
+#include "propagateremotemove.h"
+#include "propagateremotedelete.h"
+#include "common/asserts.h"
+#include "configfile.h"
+
+#include <QNetworkAccessManager>
+#include <QFileInfo>
+#include <QDir>
+#include <cmath>
+#include <cstring>
+
+#ifdef Q_OS_UNIX
+#include <unistd.h>
+#include <arpa/inet.h>
+#endif
+
+namespace OCC {
+
+Q_LOGGING_CATEGORY(lcZsyncSeed, "sync.propagate.zsync.seed", QtInfoMsg)
+Q_LOGGING_CATEGORY(lcZsyncGenerate, "sync.propagate.zsync.generate", QtInfoMsg)
+Q_LOGGING_CATEGORY(lcZsyncGet, "sync.networkjob.zsync.get", QtInfoMsg)
+Q_LOGGING_CATEGORY(lcZsyncPut, "sync.networkjob.zsync.put", QtInfoMsg)
+
+bool isZsyncPropagationEnabled(OwncloudPropagator *propagator, SyncFileItemPtr item)
+{
+    ConfigFile cfg;
+    if (!propagator->account()->capabilities().zsyncSupported()) {
+        qCInfo(lcPropagator) << "[zsync disabled] Lack of server support.";
+        item->_isZsync = false;
+        return false;
+    }
+    if (item->_remotePerm.hasPermission(RemotePermissions::IsMounted) ||
+        item->_remotePerm.hasPermission(RemotePermissions::IsMountedSub)) {
+        qCInfo(lcPropagator) << "[zsync disabled] External storage not supported.";
+        item->_isZsync = false;
+        return false;
+    }
+    if (!(cfg.deltaSyncEnabled() ||
+          propagator->_deltaSyncEnabled.fetchAndAddAcquire(0))) {
+        qCInfo(lcPropagator) << "[zsync disabled] Client configuration option.";
+        item->_isZsync = false;
+        return false;
+    }
+    if (item->_size < cfg.deltaSyncMinFileSize() ||
+        item->_size < propagator->_deltaSyncMinFileSize.fetchAndAddAcquire(0)) {
+        qCInfo(lcPropagator) << "[zsync disabled] File size is smaller than minimum.";
+        item->_isZsync = false;
+        return false;
+    }
+
+    item->_isZsync = true;
+    return true;
+}
+
+void ZsyncSeedJob::seed()
+{
+    // Create a temporary file to use with zsync_begin()
+    QTemporaryFile zsyncControlFile;
+    zsyncControlFile.open();
+    zsyncControlFile.write(_zsyncData.constData(), _zsyncData.size());
+    zsyncControlFile.flush();
+
+    int fileHandle = zsyncControlFile.handle();
+    FILE* f = fdopen(dup(fileHandle), "r");
+    setvbuf(f, NULL, _IONBF, 0);
+    rewind(f);
+
+    struct zsync_state *zs;
+    if ((zs = zsync_begin(f)) == NULL) {
+        QString errorString = tr("Unable to parse zsync file.");
+        emit failedSignal(errorString);
+        return;
```

Comment:
Fixed the closing of `f`. Duplicating the handle was because I read that closing a file handle for a QFile is a bug: https://stackoverflow.com/questions/9465727/convert-qfile-to-file/16876527#16876527

In my case `QTemporaryFile` will only get destroyed once this function ends.

Reply:
It's gone too, check above.